contracts/zero-ex/contracts/src/features/MultiplexFeature.sol

```diff
@@ -0,0 +1,799 @@
+            );
+            pairAddress = to;
+            sellAmount = outputTokenAmount;
+        }
+    }
+
+    // Same as the LiquidityProviderFeature, but without the transfer in
+    // (which is potentially done in the previous hop of a multi-hop fill)
+    // and without the minBuyAmount check (which is performed at the top, i.e.
+    // in either `batchFill` or `multiHopFill`).
+    function _sellToLiquidityProvider(
+        IERC20TokenV06 inputToken,
+        IERC20TokenV06 outputToken,
+        uint256 inputTokenAmount,
+        ILiquidityProvider provider,
+        address recipient,
+        bytes memory auxiliaryData
+    )
+        public
+        returns (uint256 outputTokenAmount)
+    {
+        uint256 balanceBefore = IERC20TokenV06(outputToken).getTokenBalanceOf(recipient);
+        if (IERC20TokenV06(inputToken).isTokenETH()) {
+            sandbox.executeSellEthForToken(
+                provider,
+                outputToken,
+                recipient,
+                0,
+                auxiliaryData
+            );
+        } else if (IERC20TokenV06(outputToken).isTokenETH()) {
+            sandbox.executeSellTokenForEth(
+                provider,
+                inputToken,
+                recipient,
+                0,
+                auxiliaryData
+            );
+        } else {
+            sandbox.executeSellTokenForToken(
+                provider,
+                inputToken,
+                outputToken,
+                recipient,
+                0,
+                auxiliaryData
+            );
+        }
+        outputTokenAmount = IERC20TokenV06(outputToken).getTokenBalanceOf(recipient)
+            .safeSub(balanceBefore);
+        emit LiquidityProviderSwap(
+            address(inputToken),
+            address(outputToken),
+            inputTokenAmount,
+            outputTokenAmount,
+            address(provider),
+            recipient
+        );
+        return outputTokenAmount;
+    }
+
+    // Some liquidity sources (e.g. Uniswap, Sushiswap, and PLP) can be passed
+    // a `recipient` parameter so the boguht tokens are transferred to the
+    // `recipient` address rather than `msg.sender`.
+    // Some liquidity sources (also Uniswap, Sushiswap, and PLP incidentally)
+    // support a "transfer then execute" pattern, where the token being sold
+    // can be transferred into the contract before calling a swap function to
+    // execute the trade.
+    // If the current hop in a multi-hop fill satisfies the first condition,
+    // and the next hop satisfies the second condition, the tokens bought
+    // in the current hop can be directly sent to the target contract of
+    // the next hop to save a transfer.
+    function _computeHopRecipient(
+        WrappedMultiHopCall[] memory calls,
+        uint256 i
+    )
+        private
+        view
+        returns (address recipient)
+    {
+        recipient = msg.sender;
+        if (i != calls.length - 1) {
```

Comment:
Given that the iterator, `i`, is computed elsewhere I might strictly enforce that `i < calls.length`.

Reply:
That said, new versions of Solidity may revert on invalid indices for `calls`. Not totally sure.